### PR TITLE
PR: Add results in batches and limit results on Find pane to a maximum

### LINKED ISSUE #10264: Find in files freezes spyder
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
Find in files will freeze spyder (alsmost completely not resonsive) if there is too many results.

The profiles tells me that most of the time is spent in `spyder.plugins.findinfiles.ItemDelegate.sizeHint`

### What steps reproduce the problem?

1. Change cwd to `/usr/local/lib/python3.7/site-packages` (or whatever the site-packages folder is)
2. Search for `in`
3. Try to use spyder while the search takes place.

### What is the expected output? What do you see instead?

Spyder becomes almost completely unresponsive.

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by goanpeca**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

finder

- The find in files now adds the files in batches that increase over time in power of 2
2^1, 2^2 etc.... up to a max of 9 (512 items per load batch)
- The maximum is now configurable, default is 1000
- The stop button and search button now act as "one single button"
- Enter now toggles between searching and stoping (instead of always restart a search.
- The regex section ins hidden by default

### Issue(s) Resolved

Fixes #10264

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @goanpeca

### CODE REVIEW

**Comment by goanpeca** [comment]
@impact27 what do you think a reasonable limit is?

@ccordoba12 thoughts on this?

**Comment by ccordoba12** [comment]
> a maximum limit to results after which a message is displayed and the search is stopped

If the changes you've made here prevent Spyder to freeze, why do we need to stop searching at some point and show a message?

**Comment by goanpeca** [comment]
> If the changes you've made here prevent Spyder to freeze,

They do not prevent it, it just makes it less terrible. But as the list keeps growing and growing it will still become unusable so we really need to set a hard limit.

**Comment by impact27** [comment]
> > If the changes you've made here prevent Spyder to freeze,
> 
> It prevents the deadlock freeze, but as the list keeps growing Spyder is not very responsive since every update of the list (even in batch) blocks the GUI (as you can see on the animation, the spinner stops for a moment every time the batch is loaded into the tree).
> 
> Thoughts @impact27 could you test this?

Is it possible to do part of the processing in a thread? I guess only GUI operations need to be done in the main thread. What part of the processing actually takes time?

**Comment by impact27** [comment]
> @impact27 what do you think a reasonable limit is?

I am not sure I would want to go through more than 1000 results. This could be a setting that is initially at 1000.

**Comment by impact27** [comment]
For reference, searchin for `in` took 15 min:
 - 10.5 min in `spyder/plugins/findinfiles/widgets.py::ItemDelegate::sizeHint`:
    - 6 min in `idealWidth`
    - 3.5 min in `setHtml`
 - 3 min in `spyder/plugins/findinfiles/widgets.py::FindInFilesWidget::search_complete`:
    - 3 min in `expandAll`
 - 1.5 min in `spyder/plugins/findinfiles/widgets.py:: ResultsBrowser::append_result`

**Comment by goanpeca** [comment]
@impact27 

>For reference, searching for in took 15 min:

Are these values with this PR? or with master?

**Comment by impact27** [comment]
Oups you are right somehow I didn’t checkout the correct branch. So these values are for the master.

**Comment by goanpeca** [comment]
>Is it possible to do part of the processing in a thread? I guess only GUI operations need to be done in the main thread. What part of the processing actually takes time?

The search is donde in a thread, I will try to move everything except the actual loading of the tree and see how it looks.

**Comment by goanpeca** [comment]
Could you provide some numbers for this branch :-) ?

**Comment by impact27** [comment]
I’ll try to do this tomorrow

**Comment by impact27** [comment]
The interface is unresposnsive for ~10s (which is much better) I ran 3 searches:
 - 18 sec in `spyder/plugins/findinfiles/widgets.py:: ResultsBrowser::append_result`:
    - 10 sec in `__init__::__repr__::get_font` (in spyder/config/gui.py)
 - 5 sec in `paintEvent` (spyder/plugins/editor/widgets/codeeditor.py)

**Comment by impact27** [comment]
Maybe `get_font` could be cached?

**Comment by goanpeca** [comment]
@impact27 I incorporated your suggestion and added some extra stuff. Please see the Animation on the description.

If you could test this, it would be great!

**Comment by ccordoba12** [comment]
> Please see the Animation on the description.

Things look pretty slick now @goanpeca! Thanks a lot for working on this!

**Comment by goanpeca** [comment]
I still need to add 1 test , working on it now.

**Comment by goanpeca** [comment]
This one is ready @ccordoba12 and @impact27 for review and final testing!

**Comment by impact27** [comment]
This totally solves the issue for me.

File: spyder/plugins/findinfiles/plugin.py [hunk 1]
```diff
@@ -53,24 +53,30 @@ def __init__(self, parent=None):
         case_sensitive = self.get_option('case_sensitive')
         path_history = self.get_option('path_history', [])
         search_in_index = self.get_option('search_in_index', default=0)
+        max_results = self.get_option('max_results', default=1000)
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/plugin.py]: Please add this option to our DEFAULTS dict in `config/main.py` and remove the `default=1000` from here.

File: spyder/plugins/findinfiles/widgets.py [hunk 2]
```diff
@@ -181,40 +196,157 @@ def find_string_in_file(self, fname):
...
                             if found > -1:
                                 break
+
         except IOError as xxx_todo_changeme:
             (_errno, _strerror) = xxx_todo_changeme.args
             self.error_flag = _("permission denied errors were encountered")
+
         self.completed = True
 
+    def process_results(self):
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/widgets.py]: Please add a docstring to this method to understand what it does.

File: spyder/plugins/findinfiles/widgets.py [hunk 3]
```diff
@@ -181,40 +196,157 @@ def find_string_in_file(self, fname):
...
+            if nb_files > 1:
+                text_files += 's'
+            text = "%d %s %d %s" % (num_matches, text_matches,
+                                    nb_files, text_files)
+        title = title + text
+
+        self.partial_results = []
+        self.sig_line_match.emit(items, title)
+
+    def truncate_result(self, line, start, end):
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles/widgets.py]: Please add a docstring to this too.

File: spyder/plugins/findinfiles/plugin.py (hunk 1 above)
Comment (goanpeca) [reply spyder/plugins/findinfiles/plugin.py]: Ok

File: spyder/plugins/findinfiles/widgets.py (hunk 2 above)
Comment (goanpeca) [reply spyder/plugins/findinfiles/widgets.py]: Done

File: spyder/plugins/findinfiles/widgets.py (hunk 3 above)
Comment (goanpeca) [reply spyder/plugins/findinfiles/widgets.py]: Done

File: spyder/config/main.py [hunk 4]
```diff
@@ -516,6 +516,10 @@
               'show_installation_dialog': True,
               'show_onboarding': True,
              }),
+            ('find_in_files',
+             {
+              'max_results': 1000,
+              }),
```
Comment (ccordoba12) [inline spyder/config/main.py]: `find_in_files` already has its own section, so please add this entry there.

Also, please change the `more_options` of `find_in_files` to False, to hide the exclusion options by default from our config.

Comment (goanpeca) [reply spyder/config/main.py]: Fixed (I actually had added the option in the wrong place)

File: spyder/config/main.py [hunk 5]
```diff
@@ -288,8 +288,9 @@
               'search_text_regexp': False,
               'search_text': [''],
               'search_text_samples': [TASKS_PATTERN],
-              'more_options': True,
+              'more_options': False,
```
Comment (ccordoba12) [inline spyder/config/main.py]: You forgot to bump CONF_VERSION to account for this change.

Comment (goanpeca) [reply spyder/config/main.py]: 🤦‍♂

Comment (goanpeca) [reply spyder/config/main.py]: Fixed

**Comment by jnsebgosselin** [comment]
Thanks a lot @goanpeca for working on this! Find in files is one of my favorite feature in Spyder actually, so I'm glad it got an improvement.

Would you guys mind if I open a PR to reduce a little bit the vertical space between the items, because I think they are a little bit too spaced now after this PR. 

image

**Comment by goanpeca** [comment]
>Would you guys mind if I open a PR to reduce a little bit the vertical space between the items, because I think they are a little bit too spaced now after this PR.

@jnsebgosselin that had to be done like that for performance reasons. We can set the same height for all items which avoids recalculating heights everytime. For big searches the time spend  on sizeHint is too much.

**Comment by jnsebgosselin** [comment]
Cool! I hope you find a workaround for this, thanks.

**Comment by goanpeca** [comment]
>Cool! I hope you find a workaround for this, thanks.

@jnsebgosselin so I tried and this method is a bottleneck, so Spyder becomes completely unresponsive if the amount of results is big (this needs a better definition).

So the workaround is still calling for every item even if I cache the heigh values with the first entries.

The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.

Thoughts @ccordoba12 ?

**Comment by jnsebgosselin** [comment]
Ok thanks for looking into it. I will try to take a look at it in case I can find a solution.

**Comment by ccordoba12** [comment]
> The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.

Sounds good to me.

**Comment by jnsebgosselin** [comment]
> > The only workaround would be to us my "hack" for maximum results bellow... 1000, but for anything above that value, go back to fixed heights.
> 
> Sounds good to me.

Ok, I've looked into this a little bit and yeah... considering the amount of data involved sometimes, I think the only solution to fix this would be to use a QTreeView instead of a QTreeWidget, but that may require a lot of work...

So like @ccordoba12, I agree with the workaround you are proposing too.

**Comment by jnsebgosselin** [comment]
What if we go with a GUI design similar to the file switcher? That would solve the variable lines height issue while not altering the performances. Also, this doesn't look bad at all...

image

**Comment by goanpeca** [comment]
Fine by me. Thoughts @ccordoba12 ?

**Comment by ccordoba12** [comment]
I like it too. Please go ahead with that idea @jnsebgosselin!

**Pull request merged**